### PR TITLE
feat(gq): undirected edge traversal — `$a <edge> $b`

### PR DESCRIPTION
**Pull request by aaltshuler**
Implements `iss-gq-undirected-traversal` via the cheap Expand-internal design — **no dependency on iss-744/579's Union machinery** (direction is handled inside the operator; alternation will compose later by carrying the flag per branch).

## What ships

`$a <edgeName> $b` matches the edge in **either direction**, with set semantics (both-ways pairs and self-loops appear once). Composes with hop bounds (`<knows>{1,3}`) and `not { }` ("no edge in either direction"). Only valid on same-endpoint-type edges — an asymmetric edge (e.g. `WorksAt: Person -> Company`) is rejected at typecheck with the new **T22** (it's well-typed in at most one orientation; the error points at the directional form).

## Why it's cheap

The machinery already existed: the IR's `Direction` enum gains `Both`; the topology index always builds both CSR and CSC; the indexed arm already had the reverse probe (`("dst","src")`); the per-source dedup gates already guard emission on both arms. The angle-bracket token is collision-free (comparisons live in the structurally separate `filter` production).

## A real bug found red-first

The undirected anti-join test failed against the naive implementation and exposed a lowering gap: negation inners are typechecked into a **discarded context clone**, so the direction lookup silently fell back to `Out` inside `not { }`. Fix: undirectedness travels on the AST node itself — the syntax is the source of truth, immune to context-propagation gaps.

## Verification

- Compiler: parse (bare/bounded/inside-not), typecheck (Both on `Knows`, T22 on `WorksAt`), 254 tests green.
- Engine: `traversal.rs` (directional-miss demonstration + out ∪ in, both-ways dedup, var-length from an incoming-only node, undirected vs directional anti-join), `traversal_indexed.rs` CSR==indexed, and `proptest_equivalence.rs` widened with `<knows>{1,3}` — CSR == indexed == auto over generated graphs.
- `cargo test --workspace --locked` green (68 suites).

Motivating case: the dev graph's `issue_related` dashboards silently missed incoming `IssueRelated` edges; with this, one undirected pattern replaces the two-query workaround.

---

<a href="https://app.devin.ai/review/modernrelay/omnigraph/pull/336" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

<h3>Greptile Summary</h3>

Adds `$a <edge> $b` undirected edge traversal syntax to GQ, matching a same-endpoint-type edge in either direction with set semantics (both-ways pairs and self-loops appear once). The implementation extends the existing `Direction` enum with a `Both` variant and threads it through parser → typecheck → IR lowering → CSR/indexed execution paths, guarded by a new T22 typecheck error for asymmetric edges.

- **Parser/typecheck/IR**: `undirected_edge = { "<" ~ edge_ident ~ ">" }` is parsed into `Traversal::undirected: bool`; typecheck enforces T22 (same-endpoint-type only) then overrides the inferred direction to `Both`; lowering carries direction from the AST node directly (bypassing the `ResolvedTraversal` context lookup that was discarded for negation inners — the root-cause fix described in the PR).
- **Engine (CSR)**: `execute_expand_csr` additionally walks the CSC (`adj_rev`) per source node, chained with the CSR scan; the existing `visited`/`seen_dst_dense` gates provide correct dedup for bidirectional pairs and self-loops.
- **Engine (indexed)**: `execute_expand_indexed` iterates `endpoint_probes(Both) = [("src","dst"),("dst","src")]`, merging both BTREE scans into a single `neighbor_map` per hop; coverage is priced conservatively as the worst of both probed columns via the new `worse_coverage` helper; the anti-join mask checks both CSR and CSC adjacency for "no edge in either direction".

<h3>Confidence Score: 5/5</h3>

Safe to merge. The change adds a new syntax form with a correctness guard (T22) at typecheck, extends existing well-tested BFS machinery, and includes comprehensive tests across parser, typecheck, lowering, CSR, indexed, and property-based paths.

All three execution arms (CSR, indexed, auto) are covered by tests, including the dedup edge cases (bidirectional pairs, self-loops) and the negation-inner regression that motivated carrying direction on the AST node. The T22 guard prevents asymmetric-edge misuse at compile time. No existing behavior is altered for directional traversals.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| crates/omnigraph/src/exec/query.rs | Core execution changes: adds endpoint_probes, worse_coverage, and adj_rev/CSC walk in both the CSR and indexed expand paths; anti-join mask updated for both-direction check. Direction grouping in both expand arms is consistent. |
| crates/omnigraph-compiler/src/query/typecheck.rs | T22 guard added before binding analysis; direction is overridden to Both after inference (no-op for same-type edges); direction correctly declared mut. Clean design. |
| crates/omnigraph-compiler/src/ir/lower.rs | Direction resolved from AST traversal.undirected flag instead of the context lookup (the root-cause fix for the negation-inner regression); dst_type and src_type for Both correctly use from_type per T22 guarantee. |
| crates/omnigraph-compiler/src/query/query.pest | New undirected_edge rule; grammar comment correctly notes angle-bracket is collision-free with comparison operators in the structurally separate filter production. |
| crates/omnigraph/tests/traversal.rs | Three new tests cover: out-union-in union + dedup, variable-length hops from incoming-only nodes, undirected vs directional anti-join. |
| crates/omnigraph/tests/traversal_indexed.rs | One-hop undirected CSR==indexed test plus a degraded-coverage test that appends an unindexed fragment to exercise the pessimistic coverage path. |
| crates/omnigraph/tests/proptest_equivalence.rs | Adds related query with <knows>{1,3} to the property-test suite, widening the CSR==indexed==auto invariant to cover undirected multi-hop over generated graphs. |
| crates/omnigraph-compiler/src/types.rs | Adds Direction::Both variant with docstring; no serialization or wire-format exposure. |
| docs/user/queries/index.md | Undirected traversal syntax documented inline with T22, hop-bounds, and not{} composition. Cost note is accurate user-facing guidance. |

<h3>Flowchart</h3>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A["a edge b undirected syntax"] --> B["Parser: Traversal undirected=true"]
    B --> C{Typecheck}
    C -->|"from_type != to_type"| D["T22 error asymmetric edge rejected"]
    C -->|"from_type == to_type"| E["direction = Both ResolvedTraversal pushed"]
    E --> F["IR Lowering reads traversal.undirected not context lookup"]
    F --> G["IROp::Expand direction=Both"]
    G --> H{"execute_expand coverage = worse of src_col and dst_col"}
    H -->|indexed path| J["execute_expand_indexed probes src+dst BTREE per hop neighbor_map merges both seen_dst dedup"]
    H -->|CSR path| K["execute_expand_csr chain adj CSR + adj_rev CSC per frontier node visited + seen_dst dedup"]
    G --> L["IROp::AntiJoin not block"]
    L --> M["try_bulk_anti_join_mask checks both adj and adj_rev has_neighbors"]
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
flowchart TD
    A["a edge b undirected syntax"] --> B["Parser: Traversal undirected=true"]
    B --> C{Typecheck}
    C -->|"from_type != to_type"| D["T22 error asymmetric edge rejected"]
    C -->|"from_type == to_type"| E["direction = Both ResolvedTraversal pushed"]
    E --> F["IR Lowering reads traversal.undirected not context lookup"]
    F --> G["IROp::Expand direction=Both"]
    G --> H{"execute_expand coverage = worse of src_col and dst_col"}
    H -->|indexed path| J["execute_expand_indexed probes src+dst BTREE per hop neighbor_map merges both seen_dst dedup"]
    H -->|CSR path| K["execute_expand_csr chain adj CSR + adj_rev CSC per frontier node visited + seen_dst dedup"]
    G --> L["IROp::AntiJoin not block"]
    L --> M["try_bulk_anti_join_mask checks both adj and adj_rev has_neighbors"]
```

</a>

<details><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `crates/omnigraph/src/exec/query.rs`, line 1050-1066 ([link](https://github.com/modernrelay/omnigraph/blob/742c8aa84519aedfbf8cf91806c529a46307d58f/crates/omnigraph/src/exec/query.rs#L1050-L1066)) 

   <a href="#">P2</a> **Cost model underestimates `Direction::Both` work by ~2×**

   `gather_cost_inputs` returns a single `edge_count` and a `src_node_count` derived from `from_type` for `Direction::Both`, but the indexed path actually fires two BTREE scans per hop — once keyed by `src` and once by `dst` (see `endpoint_probes`). The planner's `~hops × frontier × fanout` cost estimate is therefore half the real indexed work for undirected traversals. This can tip the planner toward the indexed path when CSR would be cheaper for large frontiers or densely connected undirected edges.

   The user-facing doc update already notes "its cost is roughly twice the directional equivalent," but the cost model feeding the decision doesn't reflect this. A simple correction is to double `edge_count` (or the derived `effective_max_hops`-weighted product) when `direction == Direction::Both`.

   Note: If this suggestion doesn't match your team's coding style, reply to this and let me know. I'll remember it for next time!

   <a href="https://app.greptile.com/ide/claude-code?prompt=This%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20crates%2Fomnigraph%2Fsrc%2Fexec%2Fquery.rs%0ALine%3A%201050-1066%0A%0AComment%3A%0A**Cost%20model%20underestimates%20%60Direction%3A%3ABoth%60%20work%20by%20~2%C3%97**%0A%0A%60gather_cost_inputs%60%20returns%20a%20single%20%60edge_count%60%20and%20a%20%60src_node_count%60%20derived%20from%20%60from_type%60%20for%20%60Direction%3A%3ABoth%60%2C%20but%20the%20indexed%20path%20actually%20fires%20two%20BTREE%20scans%20per%20hop%20%E2%80%94%20once%20keyed%20by%20%60src%60%20and%20once%20by%20%60dst%60%20%28see%20%60endpoint_probes%60%29.%20The%20planner's%20%60~hops%20%C3%97%20frontier%20%C3%97%20fanout%60%20cost%20estimate%20is%20therefore%20half%20the%20real%20indexed%20work%20for%20undirected%20traversals.%20This%20can%20tip%20the%20planner%20toward%20the%20indexed%20path%20when%20CSR%20would%20be%20cheaper%20for%20large%20frontiers%20or%20densely%20connected%20undirected%20edges.%0A%0AThe%20user-facing%20doc%20update%20already%20notes%20%22its%20cost%20is%20roughly%20twice%20the%20directional%20equivalent%2C%22%20but%20the%20cost%20model%20feeding%20the%20decision%20doesn't%20reflect%20this.%20A%20simple%20correction%20is%20to%20double%20%60edge_count%60%20%28or%20the%20derived%20%60effective_max_hops%60-weighted%20product%29%20when%20%60direction%20%3D%3D%20Direction%3A%3ABoth%60.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=modernrelay%2Fomnigraph&pr=336&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaudeDark.svg?v=6"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaude.svg?v=6">Fix in Claude Code</picture></a>
</details>

<sub>Reviews (3): Last reviewed commit: ["test: lowering-boundary pins for Both + ..."](https://github.com/modernrelay/omnigraph/commit/e8bba83e779a0da00038eb857fb76adbf8de8ab9) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=41942284)</sub>

### CODE REVIEW

File: crates/omnigraph/src/exec/query.rs
[the remark this answers is not in the document]
Comment (aaltshuler) [reply crates/omnigraph/src/exec/query.rs:1]: Fixed in 7591d69b — coverage is now the pessimistic combination over every probed column (Degraded dominates), and Both groups with Out in both anti-join arms with a comment pinning the cross-dependency.

File: crates/omnigraph/src/exec/query.rs [hunk 1]
```diff
@@ -1735,12 +1771,17 @@ fn try_bulk_anti_join_mask(
...
     let adj = match direction {
-        Direction::Out => gi.csr(edge_type),
+        Direction::Out | Direction::Both => gi.csr(edge_type),
         Direction::In => gi.csc(edge_type),
     }?;
+    // Undirected anti-join: "no edge in EITHER direction".
+    let adj_rev = match direction {
+        Direction::Both => Some(gi.csc(edge_type)?),
+        _ => None,
+    };
```
[the remark this answers is not in the document]
Comment (aaltshuler) [reply crates/omnigraph/src/exec/query.rs:1814]: Fixed in 7591d69b — coverage is now the pessimistic combination over every probed column (Degraded dominates), and Both groups with Out in both anti-join arms with a comment pinning the cross-dependency.
